projects/pgai/tests/vectorizer/test_vectorizer_cli.py

```diff
@@ -1035,3 +1047,112 @@ def test_vectorization_successful_with_null_contents(
+            "formatting_python_template('$chunk')",
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    "embedding",
+    [
+        ("openai/text-embedding-3-small", 1536, {}, "OPENAI_API_KEY"),
+        ("voyage/voyage-3-lite", 512, {}, "VOYAGE_API_KEY"),
+        ("mistral/mistral-embed", 1024, {}, "MISTRAL_API_KEY"),
+        ("cohere/embed-english-v3.0", 1024, {}, "COHERE_API_KEY"),
+        (
+            "huggingface/microsoft/codebert-base",
+            768,
+            {"wait_for_model": True, "use_cache": False},
+            "HUGGINGFACE_API_KEY",
+        ),
+        (
+            "azure/text-embedding-3-small",
+            1536,
+            {"api_base": os.environ["AZURE_OPENAI_API_BASE"], "api_version": os.environ["AZURE_OPENAI_API_VERSION"]},
+            "AZURE_OPENAI_API_KEY",
+        ),
+    ],
+)
+def test_litellm_vectorizer(
+    source_table: str,
+    cli_db: tuple[TestDatabase, Connection],
+    cli_db_url: str,
+    embedding: tuple[str, int, dict[str, Any], str],
+    vcr_: Any,
+    test_params: tuple[int, int, int, str, str],
+):
+    model, dimensions, extra_options, api_key_name = embedding
+    function = "embedding_litellm"
+
+    if model == "huggingface/microsoft/codebert-base":
+        pytest.skip("unable to get huggingface tests to reproduce reliably")
+
+    if api_key_name not in os.environ:
+        pytest.skip(f"environment variable '{api_key_name}' unset")
+
+    embedding_str = f"{function}('{model}', {dimensions}, extra_options => '{json.dumps(extra_options)}'::jsonb)"  # noqa: E501 Line too long
+
+    vectorizer_id = configure_vectorizer(
+        source_table, cli_db, test_params, embedding_str
+    )
+
+    num_items, concurrency, batch_size, _, _ = test_params
+    _, conn = cli_db
+    # Insert pre-existing embedding for first item
+    with conn.cursor() as cur:
+        cur.execute(
+            sql.SQL("""
+           INSERT INTO
+           blog_embedding_store(embedding_uuid, id, chunk_seq, chunk, embedding)
+           VALUES (gen_random_uuid(), 1, 1, 'post_1',
+            array_fill(0, ARRAY[{0}])::vector)
+        """).format(dimensions)
+        )
+
+    stripped_model = re.sub(r"\W", "_", model)
+
+    # When running the worker with cassette matching original test params
+    cassette = f"{function}_{stripped_model}_{dimensions}_items_{num_items}_batch_size_{batch_size}.yaml"  # noqa: E501 Line too long
+    logging.getLogger("vcr").setLevel(logging.DEBUG)
+    with vcr_.use_cassette(cassette):
+        result = CliRunner().invoke(
+            vectorizer_worker,
+            [
+                "--db-url",
+                cli_db_url,
+                "--once",
+                "--vectorizer-id",
+                str(vectorizer_id),
+                "--concurrency",
+                str(concurrency),
+                "--log-level",
+                "debug",
+            ],
+            catch_exceptions=False,
+        )
+
+    if result.exception:
+        print(result.output)
```

Comment:
this is a bit annoying, I mean we set `catch_exception=False` why doesn't it just throw the original one 😅

Reply:
A couple of things:

- I didn't know what the mechanism of `catch_exceptions=False` is. It basically says "if the CLI throws an exception which is not `SystemExit`, then re-throw that exception as-is".
- We've got try/except wrapping around the internals anyway, to ensure that the worker doesn't exit if an error occurs (because it's supposed to be a long-running thing). As a result, we should never throw anything that isn't `SystemExit`.

Reply:
We don't really log the exceptions well either. I tried adding exc_info to our error log in the cli before to get the full stack trace, but somehow that makes the cli never finish, not sure what exactly the issue is there.
But the dev experience on the tests is a bit annoying I always end up running the debugger for this.